### PR TITLE
DocumentationExtraction: Ignore fancy string

### PR DESCRIPTION
**Pull request by damngamerz**
Add regex method to ignore fancy strings.
Closes https://github.com/coala/coala/issues/4291

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by NiklasMM** [comment]
Where are the tests for this?

**Comment by damngamerz** [comment]
@NiklasMM @Makman2 @SanketDG This is one of the solution. Although Niklas suggested me to change the begin regex. I found begin regex is further bound to `marker[2]` in some cases. Also it caused some serious issues with docstrings starting with `column=0`.

Hence I moved to this solution. Hope this solves it to some extent.
Also I have for now included fancy string test for python, for other languages teach me senpai xD :stuck_out_tongue_closed_eyes:

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 1]
```diff
@@ -162,14 +162,21 @@ def _extract_doc_comment(content, line, column, markers):
...
-    if markers[1] == '':
-        # Extract and align to start marker.
-        return _extract_doc_comment_simple(content, line, column, markers)
-    elif markers[1] == markers[2]:
-        # Search for the each-line marker until it runs out.
-        return _extract_doc_comment_continuous(content, line, column, markers)
-    else:
-        return _extract_doc_comment_standard(content, line, column, markers)
+    # This ignores fancy docstrings.
+    valid_docstring_regex = '\s*[' + markers[0] + '].*'
```
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: `[]` defines a character class, not a group. Use `()` instead.

Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Actually I don't understand the regex. It will only match `markers[0]` in the worst case.

Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: ah okay you use `split()`. Though this looks super complicated, can't you just modify the begin regex here: https://github.com/coala/coala/pull/4383/commits/e44c88d5b00728c12eac417ca37170c4f0a3f50e#diff-df547e1526ff7317adef884f7414cdd8R247  ?

**Comment by NiklasMM** [comment]
> Also I have for now included fancy string test for python, for other languages teach me senpai xD :stuck_out_tongue_closed_eyes:

No, your're the one getting paid for this. Changes without proper testing are not acceptable and we're not here to spoon-feed them to you. Please create explicit test cases in `DocumentationExtractionTest` that make it very clear, what we expect to happen here, just adding this one string variable ist not making anything clear e.g. for someone just looking at documentation extraction.

I don't understand, why you don't make the changes at the first position where the begin marker is found, as I described [here](https://github.com/coala/coala/pull/4383#issuecomment-310141092)

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py (hunk 1 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: @Makman2 
>[] defines a character class, not a group. Use () instead.

yup a group :O this is good.

>Actually I don't understand the regex. It will only match markers[0] in the worst case.

Ok so what im doing is that I find the valid docstrings. If you see `line` var. it contains only the starting line of the docstring. hence with `contain[line]` we get the first raw line. now regex only matches with `     """`. If theres something suppose `   abc="""xyz` on the first line. The regex returns `"""xyz` now I compare it with stripped first line(i.e `content[line]`) and voila as the strings doesn't match it is being ignored. The main reason of using `strip()` is, the regex I don't know why doesn't match the `\n` eg. if first raw line is `    """\n` regex matches `    """`.

Now I tried modifying begin regex but the problem I m facing is [here](https://github.com/damngamerz/coala/blob/e44c88d5b00728c12eac417ca37170c4f0a3f50e/coalib/bearlib/languages/documentation/DocumentationExtraction.py#L199)
regex will get spaces. Hence raising `KeyError` I can't use `strip()` as some markers have spaces in it.

**Comment by damngamerz** [comment]
@NiklasMM I already did my homework :D
Most of the languages. have explicit markers for docstrings like in C,java...
And hence their assignment of strings are always in form `abc="xyz"`. leaving no space for using those explicit markers for assignment. This is the case only in python. Although I was just asking if there's any which i might have missed.

>Please create explicit test cases in DocumentationExtractionTest

Yes I have some more tests in mind. Amending them atm.

About the begin marker let me try again if can find a workaround to the issue which I have mentioned in my [comment](https://github.com/coala/coala/pull/4383#discussion_r123685756) above.

**Comment by NiklasMM** [comment]
> And hence their assignment of strings are always in form abc="xyz". leaving no space for using those explicit markers for assignment. This is the case only in python. Although I was just asking if there's any which i might have missed.

No, this seems right. Python seems to be the weirdo in that regard ;)

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py (hunk 1 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Ok so it goes like this `marker_dict` is specifically made for `begin_regex` to find the `marker[0]` exactly as it is.
As `marker[0]` is the key of `marker_dict` which will return `markers`. eg`marker_dict` for python is `['"""':['"""','','"""']]`.
Hence, this is the reason why I can't change/tamper `begin_regex`. It would require a major change so as to do it(which I think it should not be removed as it has its importance). look at this [for more info.](https://github.com/damngamerz/coala/blob/690ab278c331ffa2c6d53179e5d010b23c9107f6/coalib/bearlib/languages/documentation/DocumentationExtraction.py#L230)

Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Further modified regex :tada: got this idea from `begin_regex` no need of group matching `()`. we now add escape sequence before each marker character(This is required as for other languages special character needs escape sequences eg: java `*/**`). So for python the regex is now
`\s*\"|\"|\".*`

**Comment by damngamerz** [comment]
With the help of @NiklasMM I was able to amend the changes through `begin_regex` :tada: 
Hope this solves it.
CC^ @Makman2

File: tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py [hunk 2]
```diff
@@ -64,3 +64,7 @@ def foobar_triangle(side_A, side_B, side_C):
     :return: returns perimeter
     """
     return side_A + side_B + side_C
+
+    fancy_string_variable = """
```
Comment (SanketDG) [inline tests/bearlib/languages/documentation/documentation_extraction_testdata/default.py]: can you add a comment this is being ignored

**Comment by SanketDG** [comment]
ack e3056fd

**Comment by SanketDG** [comment]
I'll let Mischa take one final look

also, needs rebase

**Comment by SanketDG** [comment]
Also, let's not call it fancy string, lets call something like triple quote string literals

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 3]
```diff
@@ -179,23 +179,25 @@ def _compile_multi_match_regex(strings):
     :param strings: The strings to match.
     :return:        A regex object.
     """
-    return re.compile('|'.join(re.escape(s) for s in strings))
+    return re.compile('^\s*(?P<marker>' +
+                      ('|'.join(re.escape(s) for s in strings)) +
+                      ')')
```
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: I like the parent name `_compile_multi_match_regex` as we are adding escape sequences over each character of the docstring. Let me write a description inside the docstring :+1:

Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: Instead of compiling, we maybe just return the regex string? Then this function can still keep it's old pure functionality, and below at `begin_regex = ...` we compile it ourselves, adding the space-matching-stuff. What do you think?

Though like this it's also fine, I mean space matching is still understandable for this function :) Just an idea.

File: coalib/bearlib/languages/documentation/DocumentationExtraction.py [hunk 4]
```diff
@@ -174,28 +174,32 @@ def _extract_doc_comment(content, line, column, markers):
...
 
 def _extract_doc_comment_from_line(content, line, column, regex,
                                    marker_dict, docstyle_definition):
     cur_line = content[line]
     begin_match = regex.search(cur_line, column)
     if begin_match:
-        indent = cur_line[:begin_match.start()]
         column = begin_match.end()
-        for marker in marker_dict[begin_match.group()]:
+        indent = cur_line[:column-len(begin_match.group('marker'))]
```
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: I would rather match the complete indent and reuse it here :+1:
```
begin_match.group(1)  # <-- would instead contain the full indentation
# we could optionally assert that begin_match.start() == 0
```

Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationExtraction.py]: yup i like the idea of group matching `indent` too and use it here to extract indent :+1:

**Comment by Makman2** [comment]
ack 89b39dc

**Comment by SanketDG** [comment]
rebase needed

**Comment by damngamerz** [comment]
@SanketDG done :+1:

**Comment by SanketDG** [comment]
reack 0114024

**Comment by SanketDG** [comment]
@rultor merge
